# Grupo 1
### \[1.5 + 1.5 + 2.0 + 2.0 valores\] Defina as seguintes funções:

1. `digitos`, que recebe um número inteiro positivo e devove uma lista de inteiros, composta pelos seus dígitos.

    Por exemplo,
    
    `digitos(538) = [5, 3, 8]`

In [5]:
def digitos(n):
    digs = []
    while n != 0:
        digs.insert(0, n%10)
        n = n // 10
    return digs

digitos(538)

[5, 3, 8]

2. `dividirPor3`, que determina se um valor inteiro positivo é divisível por três, usando o algoritmo de somar os seus dígitos e verificar se esta soma é divisível por três.

    Por exemplo,

    `dividirPor3(538) = False` porque `5+3+8=16` e `16` não é divisível por três.

In [7]:
def dividirPor3(n):
    soma = sum(digitos(n))
    return soma % 3 == 0

dividirPor3(538)

False

3. `dividirPor3`, rescreva a função anterior de forma que, se a soma dos dígitos for maior ou igual a 10, então deve continuar a aplicar o mesmo algoritmo recursivamente até chegar a um valor menor que 10. Aí, se o valor final for 3, 6, ou 9, então o número original é divisível por três.

    Por exemplo,

    `dividirPor3(538) = dividirPor3(16) = dividirPor3(7) = False`

In [9]:
def dividirPor3(n):
    soma = sum(digitos(n))

    if soma >= 10:
        return dividirPor3(soma)
    elif soma in (3, 6, 9):
        return True
    else:
        return False
dividirPor3(538)

False

4. Considere a função `somaDois` que recebe um inteiro positivo *n* e calcula a soma dos *n* primeiros termos da sequência **2,22,222,2222....**

    Por exemplo,

    `somaDois(0) = 0`   `somaDois(3) = 2+22+222 = 246`

    Implemente `somaDois` com listas por compreensão (pode usar outras funções para além das listas).

    Se não conseguir fazer com listas por compreensão, resolva de outra forma para metade da cotação.

In [15]:
def somaDois(n):
    seq = [int('2'*(i+1)) for i in range(n)]
    return sum(seq)

print('somaDois(0)', somaDois(0))
print('somaDois(3)', somaDois(3))

somaDois(0) 0
somaDois(3) 246


# Grupo 2
### \[1.0 + 1.0 + 1.0 valores\] Seja a função fatorial definida da seguinte forma:

In [24]:
def fatorial(n):
    if n == 0:
        return 1
    return n * fatorial(n-1)

1. Inclua duas pré-condições necessárias.

In [ ]:
def fatorial(n):
    """
    Requires: type(n) == int and n>=0
    """
    if n == 0:
        return 1
    return n * fatorial(n-1)

2. Inclua dois testes relevantes usando a sintaxe da biblioteca `doctest`.

In [ ]:
def fatorial(n):
    """
    >>> fatorial(0)
    1
    >>> fatorial(3)
    2
    """
    if n == 0:
        return 1
    return n * fatorial(n-1)

3. Rescreva a função usando um ciclo `for`.

In [25]:
def fatorial(n):
    prod = 1
    for i in range(1, n+1):
        prod *= i
    return prod

# Grupo 3
### \[0.75 + 0.5 + 0.75\] O tipo dicionário é uma das estruturas de dados disponívels na linguagem Python

1. Quais das seguintes características são parte dos dicionários? Justifique as suas respostas
    - mutabilidade
    - indexação
    - ordenação (i.e., existe uma ordenação dos seus elementos?)

Um dicionário é mutável pois é possível adicionar pares chave/valor, remover, e modificar (apenas o valor. Modificar a chave é remover a entrada e adicionar uma nova).

Um dicionário é também indexável através das chaves (do par chave/valor).

Um dicionário, a partir da versão 3.7, preserva a ordem de insersão, logo é ordenado (ver [isto](https://mail.python.org/pipermail/python-dev/2017-December/151283.html)).

2. Que resultado é impresso pela execução do seguinte programa? Justifique.

In [26]:
def f(dic, k, v):
    if k in dic:
        dic[k] = v*2
    else:
        dic[v] = dic[v]*2

myDic = {2:10, 4:15, 10:100}
f(myDic, 4, 10)
f(myDic, 5, 10)
print( sum(myDic.values()) )

230


O resultado impresso é **230**

É de notar que a função f afeta o dicionário porque dicionários são passados por referência.

Na linha 7 (*f(myDic, 4, 10)*) o dicionário fica com o valor **20** (10, do *v*, vezes dois) onde a chave é **15** pois o valor 4, do *k*, está presente no dicionário então executa o corpo do primeiro `if`.

Na linha 8 (*f(myDic, 5, 10)*), como a chave `k`, de valor 5, não está presente no dicionário, ir-se-à usar o valor `v` como chave, i.e., entraremos no corpo do `else`. O valor da chave 10 do dicionário será então o dobro do seu valor: 100*2 = 200

No fim, imprimimos a soma dos valores: 10+20+200 = 230
